### ISSUE TITLE
Type is not narrowed in walrus conditional

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#474]
**Describe the bug**
This is another issue that was raised on the mypy tracker that also affects Pyright.  Conditionals of assignment expressions are not narrowed for type, producing errors like this one:

```json
{
	"resource": "test.py",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "'foo' is not a known member of 'None'",
	"source": "pyright (reportOptionalMemberAccess)",
	"startLineNumber": 16,
	"startColumn": 11,
	"endLineNumber": 16,
	"endColumn": 14
}
```

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That `a` would not be `None` under the `bad` conditional.

**Screenshots or Code**
```py
# pyright: strict

from typing import Optional

class C:
    def foo(self):
        pass

def good(b: Optional[C]) -> None:
    a = b
    if a:
        a.foo()

def bad(b: Optional[C]) -> None:
    if a := b:
        a.foo()
```

**VS Code extension or command-line**
extension, 1.1.17

**Additional context**
See https://github.com/python/mypy/issues/7316.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next version of Pyright.

**Comment by erictraut** [comment]
This support is now included in 1.1.19, which I just published.

**Comment by alexkoay** [comment]
There might be a regression with this.
```
import re
def foo(s: str) -> str:
	if m := re.fullmatch('(test).+', s):
		return m.group(1)
	return 'oops'
```
`group` reports a `reportOptionalMemberAccess` error under Pyright 1.1.38.
